File: src/flink-examples/flink-examples-normal/FlinkCheckpointJavaExample/src/main/java/com/huawei/bigdata/flink/examples/SEventSourceWithChk.java
/*
 * Copyright (c) Huawei Technologies Co., Ltd. 2012-2020. All rights reserved.
 */

package com.huawei.bigdata.flink.examples;

import org.apache.flink.api.java.tuple.Tuple4;
import org.apache.flink.streaming.api.checkpoint.ListCheckpointed;
import org.apache.flink.streaming.api.functions.source.RichSourceFunction;

import java.security.SecureRandom;
import java.util.ArrayList;
import java.util.List;

/**
 * 自定义带状态Source类
 * 
 * @since 2019/9/30
 */
public class SEventSourceWithChk extends RichSourceFunction<Tuple4<Long, String, String, Integer>>
        implements ListCheckpointed<UDFState> {
    private Long count = 0L;
    private boolean isRunning = true;
    private String alphabet =
            "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWZYX0987654321";

    @Override
    public void run(SourceContext<Tuple4<Long, String, String, Integer>> ctx) throws Exception {
        SecureRandom random = new SecureRandom();
        while (isRunning) {
            for (int i = 0; i < 10000; i++) {
                ctx.collect(Tuple4.of(random.nextLong(), "hello-" + count, alphabet, 1));
                count++;
            }
            Thread.sleep(1000);
        }
    }

    @Override
    public void cancel() {
        isRunning = false;
    }

    @Override
    public List<UDFState> snapshotState(long l, long ll) throws Exception {
        UDFState udfState = new UDFState();
        List<UDFState> listState = new ArrayList<UDFState>();
        udfState.setState(count);
        listState.add(udfState);

        return listState;
    }

    @Override
    public void restoreState(List<UDFState> list) throws Exception {
        UDFState udfState = list.get(0);
        count = udfState.getState();
    }
}

File: src/flink-examples/flink-examples-normal/FlinkCheckpointScalaExample/src/main/java/com/huawei/bigdata/flink/examples/SEventSourceWithChk.scala
package com.huawei.bigdata.flink.examples

import java.util
import java.util.{ArrayList, List}

import org.apache.flink.streaming.api.checkpoint.ListCheckpointed
import org.apache.flink.streaming.api.functions.source.{RichSourceFunction, SourceFunction}

import scala.util.Random

class SEventSourceWithChk extends RichSourceFunction[SEvent] with ListCheckpointed[UDFState]{

  private var count = 0L
  private var isRunning = true

  private[examples] val listState = new util.ArrayList[UDFState]

  private val alphabet = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWZYX0987654321"

  override def run(sourceContext: SourceFunction.SourceContext[SEvent]): Unit = {
    while(isRunning) {
      for (i <- 0 until 10000) {
        sourceContext.collect(SEvent(Random.nextLong(), "hello-" + count, alphabet, 1))
        count += 1L
      }
      Thread.sleep(1000)
    }
  }

  override def cancel(): Unit = {
    isRunning = false
  }

  override def snapshotState(l: Long, l1: Long): util.List[UDFState] = {
    val udfList: util.ArrayList[UDFState] = new util.ArrayList[UDFState]()
    val udfState = new UDFState
    udfState.setState(count)
    udfList.add(udfState)
    udfList
  }

  override def restoreState(list: util.List[UDFState]): Unit = {
    if (listState.size > 0) {
      val udfState = list.get(0)
      count = udfState.getState
    }
  }

}
